packages/hash/subgraph/src/types/identifier.ts

```diff
@@ -0,0 +1,111 @@
+export type EntityVersion = string;
+
+/**
+ * An identifier of a specific edition of an `Entity` at a given `EntityVersion`
+ */
+export type EntityEditionId = {
+  baseId: EntityId;
+  version: EntityVersion;
+};
+
+/**
+ * A tuple struct of a given `EntityId` and timestamp, used to identify an `Entity` at a given moment of time, where
+ * that time may be any time in an `Entity`'s lifespan (and thus the timestamp is *not* necessarily equal to an
+ * `EntityVersion`)
+ */
+export type EntityIdAndTimestamp = {
+  baseId: EntityId;
+  timestamp: string;
+};
+
+export { OntologyTypeEditionId };
+
+export type GraphElementEditionId = EntityEditionId | OntologyTypeEditionId;
+
+export const isEntityId = (entityId: string): entityId is EntityId => {
+  const [accountId, entityUuid] = entityId.split("%");
+  return (
+    accountId != null &&
+    entityUuid != null &&
+    validateUuid(accountId) &&
+    validateUuid(entityUuid)
+  );
+};
+
+export const isEntityEditionId = (
+  editionId: object,
+): editionId is EntityEditionId => {
+  return (
+    "baseId" in editionId &&
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore-expect -- This is fixed in TS 4.9
+    typeof editionId.baseId === "string" &&
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore-expect -- This is fixed in TS 4.9
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
+    isEntityId(editionId.baseId) &&
+    "version" in editionId &&
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore-expect -- This is fixed in TS 4.9
+    typeof editionId.version === "string" &&
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore-expect -- This is fixed in TS 4.9
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
+    !Number.isNaN(Date.parse(editionId.version))
+  );
+};
+
+export const isOntologyTypeEditionId = (
+  editionId: object,
+): editionId is OntologyTypeEditionId => {
+  return (
+    "baseId" in editionId &&
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore-expect -- This is fixed in TS 4.9
+    typeof editionId.baseId === "string" &&
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore-expect -- This is fixed in TS 4.9
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
+    isEntityId(editionId.baseId) &&
+    "version" in editionId &&
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore-expect -- This is fixed in TS 4.9
+    typeof editionId.version === "number" &&
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore-expect -- This is fixed in TS 4.9
+    Number.isInteger(editionId.version) &&
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore-expect -- This is fixed in TS 4.9
+    editionId.version >= 0
```

Comment:
Can the `version` be `0`? I assumed it would start at `1`:
```suggestion
    editionId.version > 0
```

Reply:
The spec suggests it's 1, I wouldn't be surprised if we expanded it to be 0 later on to allow for `0` versions like other systems. It would be way way too easy to forget to change it here. I'm just going to remove this check because it's not needed.